### PR TITLE
add search_analyze endpoint

### PR DESCRIPTION
**Pull request by tonysun83**
This adds the search_analyze endpoint for search3
```
{"analyzer":"keyword", "text":"ablanks@renovations.com"}

{
    "tokens": [
        "ablanks@renovations.com"
    ]
}
```

### CODE REVIEW

File: src/search3_httpd.erl [hunk 1]
```diff
@@ -59,6 +59,52 @@ handle_search_req(#httpd{path_parts=[_, _, _, _, _]}=Req, _Db, _DDoc,
...
+    Text = couch_util:get_value(<<"text">>, Fields),
+    analyze(Req, Analyzer, Text);
+handle_analyze_req(Req) ->
+    send_method_not_allowed(Req, "GET,POST").
+
+analyze(Req, Analyzer, Text) ->
+    case Analyzer of
+        undefined ->
+            throw({bad_request, "analyzer parameter is mandatory"});
+        _ when is_list(Analyzer) ->
```
Comment (rnewson) [inline src/search3_httpd.erl]: `List when is_list(List)` is clearer rather than ignore the value.

File: src/search3_httpd.erl [hunk 2]
```diff
@@ -59,6 +59,52 @@ handle_search_req(#httpd{path_parts=[_, _, _, _, _]}=Req, _Db, _DDoc,
...
+            ok;
+        _ when is_binary(Analyzer) ->
+            ok;
+        {[_|_]} ->
+            ok;
+        _ ->
+            throw({bad_request, "analyzer parameter must be a string or an object"})
```
Comment (rnewson) [inline src/search3_httpd.erl:99]: we'll send this for an empty object which might be confusing.

File: src/search3_httpd.erl [hunk 3]
```diff
@@ -59,6 +59,52 @@ handle_search_req(#httpd{path_parts=[_, _, _, _, _]}=Req, _Db, _DDoc,
...
+    end,
+    case Text of
+        undefined ->
+            throw({bad_request, "text parameter is mandatory"});
+        _ when is_list(Text) ->
```
Comment (rnewson) [inline src/search3_httpd.erl]: same comment as above, but this shows the danger of this pattern. you're referring to a previously defined variable further up the function, easy to get wrong. Perhaps a few "is_string" and "is_string_or_object" functions could be pulled out for further clarity.

File: src/search3_httpd.erl [hunk 4]
```diff
@@ -59,6 +59,52 @@ handle_search_req(#httpd{path_parts=[_, _, _, _, _]}=Req, _Db, _DDoc,
...
+            ok;
+        _ when is_binary(Text) ->
+            ok;
+        _ ->
+            throw({bad_request, "text parameter must be a string"})
+    end,
+    Response = search3_rpc:analyze(couch_util:to_binary(Analyzer),
+        couch_util:to_binary(Text)),
+    couch_log:notice("Response ~p", [Response]),
```
Comment (rnewson) [inline src/search3_httpd.erl]: we don't want logs of all of these, do we?

File: src/search3_httpd.erl [hunk 5]
```diff
@@ -59,6 +59,52 @@ handle_search_req(#httpd{path_parts=[_, _, _, _, _]}=Req, _Db, _DDoc,
...
+        undefined ->
+            throw({bad_request, "text parameter is mandatory"});
+        _ when is_list(Text) ->
+            ok;
+        _ when is_binary(Text) ->
+            ok;
+        _ ->
+            throw({bad_request, "text parameter must be a string"})
+    end,
+    Response = search3_rpc:analyze(couch_util:to_binary(Analyzer),
```
Comment (rnewson) [inline src/search3_httpd.erl]: do we need to force to binary?

File: src/search3_httpd_handlers.erl [hunk 6]
```diff
@@ -6,11 +6,10 @@
...
 ]).
 
+url_handler(<<"_search_analyze">>) -> fun search3_httpd:handle_analyze_req/1;
 url_handler(_) -> no_match.
 
 design_handler(<<"_search">>) -> fun search3_httpd:handle_search_req/3;
 design_handler(_) -> no_match.
 
 db_handler(_) -> no_match.
-
```
Comment (rnewson) [inline src/search3_httpd_handlers.erl:15]: ?

**Review by rnewson — CHANGES_REQUESTED** [review]

**Comment by tonysun83** [comment]
@rnewson updated per your comments

**Review by rnewson — APPROVED** [review]
+1

**Pull request merged**